### ISSUE TITLE
Remove FlagEncoder

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#2463]
This will be a big refactoring task that we will have to split into several smaller issues. We can keep a list of related issues here and also discuss the general direction this should be going here.

* #2462

**Comment by otbutz** [comment]
IMHO #2460

**Comment by easbar** [comment]
> IMHO #2460

Yes, everything related to access and access flags is related. Just not sure if this is required to remove the FlagEncoders (or vice versa).

**Comment by otbutz** [comment]
Removing FlagEncoders requires #2460 to be solved but we can implement it without removing FlagEncoders. At least that is my understanding of it.

The benefit is that users can perform proper truck routing without touching the code :)

**Comment by easbar** [comment]
I think we are getting there :) What we probably should do next is:

* [ ] allow storing the maximum, actually used values for encoded values, so we can replace `encoder.getMaxSpeed()` with `speedEnc.getMaxValue()` or similar. This would be an improvement, because we no longer have to keep the max value synchronized between the vehicle tag parsers and the flag encoders.
* [ ] replace the `FlagEncoder` parameters from the weighting constructors with references to the encoded values they need. For example instead of receiving a `FlagEncoder` and calling `encoder.getAverageSpeedEnc()`,  the `FastestWeighting` should receive the speed encoded value (and encoded values it needs besides this of course). 
* [ ] For `FastestWeighting` we also need to replace `encoder.isMotorVehicle()/HGV()` with some boolean constructor parameters representing the same values (or just a PMap). Maybe we should take these parameters from the weighting string, i.e. something like `weighting: fastest|is_motor_vehicle=true`. Or we skip this intermediate step and replace FastestWeighting (and the others) with CustomWeighting directly.
* [ ] One problem is that we have literally thousands of usages of `GHUtility#setSpeed(..., encoder, ...)` that we will have to replace with something similar that uses the speed/access encoded values directly. I did some unfinished work in this direction some time ago here: 9715c81cef5c2eea66acfb93b72f39342dfd1ff9.

**Comment by easbar** [comment]
Not all the issues we mentioned here are fixed, but flag encoders are almost gone since #2611 :)

The only place they remain now is in `config.yml`/`Graphhopper.java`. For example `graph.flag_encoders=car,bike` does two things now: 

1) It creates `VehicleEncodedValues` for car and bike, which just means that the `car_access`, `car_average_speed`, `car_turn_cost`, `bike_access`, `bike_average_speed` and `bike_priority` encoded values are added to the graph / encoding manager.

2) It enables the `Car/BikeTagParser`s, which require above encoded values.

We also have `graph.encoded_values` which is very similar: For example `graph.encoded_values=road_environment` adds the `road_environment` encoded value and enables the `OSMRoadEnvironment` parser. The fundamental difference between `graph.flag_encoders` and `graph.encoded_values` is that the 'flag_encoders' add **multiple** encoded values via VehicleEncodedValuesFactory and a single parser via VehicleTagParserFactory that writes these encoded values, while for each entry in `graph.encoded_values`  there will be a **single** encoded value added via EncodedValuesFactory and an according single parser via TagParserFactory.

I don't think this 'minor' difference deserves an extra config option and two extra factories. Let alone it should definitely no longer be called `graph.flag_encoders`.

**Comment by easbar** [comment]
Another list taken from here: https://github.com/graphhopper/graphhopper/pull/2561#issue-1205880236:

* [ ] Remove special handling of VehicleTagParsers in ~~TagParserBundle~~OSMParsers (acceptWay method)
* [ ] Remove default encoded values from EncodingManager and default tag parsers in the GraphHopper class
* [ ] Split VehicleTagParsers into speed/access/priority etc. parsers
* [ ] Make curvatureEnc a standalone EV
* [ ] Get rid of VehicleTagParser.init(DateRangeParser)
